### DISCUSSION TITLE
HTTP Communication Issue between Podman Containers in the same Network

### DISCUSSION THREAD

**Discussion by mdsposadas** [containers/podman#17637] [Q&A]
### Introduction
The setup is for 2 Containers (Client and Server) under the same Network using Podman.
We are trying to perform inter-container communication using HTTP Requests.
The Client container is running a python script that uses urllib which tries to send a HTTP Request to the Server Container.
Current issue is that the Client container encounters the following error: **_HTTP Error 503: Service Unavailable_**

The same setup was created using Docker and it works perfectly.
Also, the same setup was used but using Socket Communication instead of HTTP communication and it works for either Podman rootful or rootless containers.
<br />

### Environment Setup (Files)
**Server**
1. index.html: file used to send as response to Client upon HTTP request
![0_nano_index_html](https://user-images.githubusercontent.com/94945936/221491343-ea532c87-074e-4756-9e36-b7755204e8cf.JPG)

2. server.py: script used to create and start the Web Server
![2_nano_server_py](https://user-images.githubusercontent.com/94945936/221491552-c3fce247-ff8a-4e1b-9760-bfdbf89afda9.JPG)

3. Dockerfile: file used to build the podman image for the Server Container
![3_nano_Dockerfile](https://user-images.githubusercontent.com/94945936/221491572-eedb94e5-9a22-47a1-bdf5-53a3241c6c1f.JPG)
<br />

**Client**
1. client.py: script used to send request to Web Server, which in turn will acquire the response from the Server
![2_nano_client_py](https://user-images.githubusercontent.com/94945936/221491991-41eef472-ba88-44ab-b4d3-1bc626342fcb.JPG)

2. Dockerfile: file used to build the podman image for the Client Container
![3_nano_Dockerfile](https://user-images.githubusercontent.com/94945936/221491999-6f7fc5f0-b140-4cad-b484-6785e960b588.JPG)
<br />

### Environment Setup (Commands)
The overall environment was created with the following commands:
1. To begin, the network is created in which the 2 containers (Client and Server) will operate.
`sudo podman network create --subnet=172.18.0.0/16 test-network`
Here is the network configuration of the network created:
![3_podman_network_inspect_test-network](https://user-images.githubusercontent.com/94945936/221492758-e4fbfecb-c254-4a0f-9866-272eaf5b14ed.JPG)

2. Afterwards, the images are built for the containers of the Server and Client.
_Server_
![4_podman_build_test_server](https://user-images.githubusercontent.com/94945936/221493091-b84ffd0b-b650-4bd3-a9b6-89f38370c38f.JPG)
_Client_
![4_podman_build_test_client](https://user-images.githubusercontent.com/94945936/221493209-15069c9e-a621-405c-b88f-f60e599e44a7.JPG)

3. The containers are then built for the Server and the Client using the Images built from the previous step.
_Server_
`sudo podman create --name=test_server --network=test-network --ip=172.18.0.2 --network-alias test_server -p 1234:1234 --restart always test_server:00.00.01 python ./server.py`
_Client_
`sudo podman create --name=test_client --network=test-network --ip=172.18.0.3 --network-alias test_client --restart always test_client:00.00.01 python ./client.py`

4. Afterwards, 2 terminals are opened to attach to each container and the following error was encountered on the Client Container:
_Server_
![6_podman_start_test_server](https://user-images.githubusercontent.com/94945936/221493714-0b3da4fd-541f-4263-b8f2-1ef3a4f44e31.JPG)
_Client_
![6_podman_start_test_client](https://user-images.githubusercontent.com/94945936/221502222-c07923f4-5cdd-4b96-ab76-6e7da9bb47c6.png)

<br />

### Results
**Expected Result**
The Client container should not encounter an error and should run smoothly, providing the message shown below which comes as the response from the Server container:
![6_podman_start_test_client_OK](https://user-images.githubusercontent.com/94945936/221494558-ea9caabc-afb4-44b9-8b4e-4929a887dc2f.JPG)

**Actual Result**
The Client container encounters the following error (**_HTTP Error 503: Service Unavailable_**) once the container starts running, as shown below:
![6_podman_start_test_client](https://user-images.githubusercontent.com/94945936/221502222-c07923f4-5cdd-4b96-ab76-6e7da9bb47c6.png)
<br />

Is there any way we can resolve this issue? This environment was also used when using _docker_ and it works smoothly.
For additional information, we are using podman v4 for this environment setup:
![218729952-81f30b88-238e-4897-9a74-4701bfc9fdda](https://user-images.githubusercontent.com/94945936/221495872-83c96e5c-dd4c-4213-a888-613b90ac74fa.png)

**Comment by rhatdan** [comment]
@Luap99 PTAL
